examples/usematches-loader-data/app/useOptionalUser.ts

```diff
@@ -0,0 +1,37 @@
+import useMatchesData from "./useMatchesData";
+
+interface User {
+  name: string;
+  email: string;
+}
+
+function isUser(user: unknown): user is User {
+  return (
+    !!user &&
+    typeof (user as User).name === "string" &&
+    typeof (user as User).email === "string"
+  );
+}
+
+/**
+ * Build a custom hook for each data object
+ * of your loader data.
+ * Use useMatchesData to access loader data
+ * across your application.
+ * Use tiny-invariant and Typescript "is"
+ * to require data on runtime.
+ * Or return undefined if data is optional and not found.
+ */
+export default function useOptionalUser(): User | undefined {
+  const { user } = useMatchesData("root");
+  /*
+   * You can make use of tiny-invariant here to throw runtime errors
+   * if the loader data is required and not optional
+   */
+  if (!isUser(user)) {
+    return undefined;
+  }
+  return user;
+}
```

Comment:
Let's call this file `user.ts` and have two named exports:

```suggestion
export function useOptionalUser(): User | undefined {
  const data = useMatchesData("root");
  /*
   * You can make use of tiny-invariant here to throw runtime errors
   * if the loader data is required and not optional
   */
  if (!data || !isUser(data.user)) {
    return undefined;
  }
  return data.user;
}

export function useUser(): User {
  const maybeUser = useUser();
  if (!maybeUser) {
    throw new Error("No user found in root loader, but user is required by useUser. If user is optional, try useOptionalUser instead.");
  }
  return maybeUser
}
```

Reply:
@kentcdodds I like it! You called `useUser` inside `useUser` though :P I fixed it and pushed the changes!